Fix scheduled update time-drift in data update coordinator

azogue opened the pull request:
In `DataUpdateCoordinator` the periodic refresh is drifting 1 second over the desired `update_interval`.

As next schedule is calculated **after** the update is done, setting now + update_interval makes 1-second drift in practice, as the HA tick is of 1Hz (so for t.0x, it will trigger in t+1, instead of t, where x is the fraction of time the update took).

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Subtract the time it took from the `update_interval` timedelta so the next schedule is not drifted. 

Apply some safety factor (1.2x, 20%) for extra time lost in operation.

### Other approaches

Another approach would be to just _floor_ `utcnow` to remove sub-second time lost in the operation, making sure that new "point_in_time" is an exact one, by doing:
`utcnow().replace(microsecond=0) + self.update_interval`

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

azogue committed:
Fix scheduled update time-drift in data update coordinator

As next schedule is calculated **after** the update is done,
setting now + update_interval makes 1 second drift in practice,
as the tick is 1Hz.

bdraco commented:
The mypy error has been fixed upstream.  Would you please resubmit this for a clean run.
https://developers.home-assistant.io/docs/development_catching_up/

elupus commented:
In general it's better to keep track of the timestamp of the last update (it gets passed to scheduled function) then use that as the base too add the delay too. Now it's using utcnow() as base which causes the drift.

azogue commented:
> In general it's better to keep track of the timestamp of the last update (it gets passed to scheduled function) then use that as the base too add the delay too. Now it's using utcnow() as base which causes the drift.

I know, but here the timestamp is not used (`async def _handle_refresh_interval(self, _now: datetime)`), and that calls `async_refresh()`, which is where the request is made (and that is a public method with many calls).

We could change the actual `start = monotonic()` for a `start = utcnow()`, but in esence is the same thing.

Also, IMHO, we cannot lose track of the main objective (for me, at least), which is to have **real periodic calls, with a constant frequency determined by its `update_interval`**. 
To reach the objective we need to understand how the infrastructure works, and what is doing the current implementation, or the change you propose:
- At `T=n + some ns/ms`, the time changed event is triggered, with the `now` timestamp on it.
- Listeners evaluate each `t_schedule >= now` condition, adding the jobs to the queue. Some time passes.
- The `_handle_refresh_interval ` is called, and that calls `async_refresh()`, running now at `T=n.xxx`.
- monotonic() start is stored to measure request time, that _in general_ would be some time < 1s, let's call it ∆c = 0.YYY.
- Finally, a new update is scheduled with `utcnow() + update_interval`. That makes the new call on `T=n + 0.xxx + ∆c + update_interval`, which will never trigger on `T=n+update_interval`, but it'll trigger on `T=n+update_interval+1` or even `T=n+update_interval+2` (depending on event bus stress and/or update request time)

You're partially right on using the `now` timestamp from the time_changed event, as that one _supposedly_ ensures the constant frequency, but be aware that the `now` timestamp is not rounded, so if millisecond precision is not removed, it will fail 50% of all calls, adding a 1 second drift. One example:
- Time changed event with `now=09:32:23.002603+00:00`, update takes 0.033 sec (but ignored now), next update is scheduled for `now+interval =09:32:28.002603+00:00`, with a period of 5s.
- In '09:32:28', even has `now=09:32:28.005022+00:00`, and activates the listener, an update is done in that tick, and period is 5s, everything fine. Next schedule now is for `2020-03-20 09:32:33.005022+00:00`
- At '09:32:33', the tick has a delay < 5 ms, **so it does not activate the listener**.
- At '09:32:34', the listener is activated again, with `now=09:32:34.001559+00:00`, and the update runs again. **But we missed a tick** and now the period was `update_interval+1`.

The sub-second delay in the `now` timestamp is always present, and it will take random values, so you can't rely on it to reach constant frequencies.

IMO, the most secure way to be sure of activation is to **_round_ the point_in_time**, which is the other, even more simpler, proposal for this issue: schedule on `utcnow().replace(microsecond=0) + self.update_interval`.

That one does also not ensure constant frequency (nor the other one, btw), but it is near it, for all calls that don't add more than 1 second with `now_delay + ∆c + queue_delay` (in a normal context it will be all of them).

balloob commented:
This is only really an issue if this object is called every second, which is not something that we allow in Home Assistant.

If the update interval is every 5 seconds (the lowest we allow) and a request takes 5 seconds to complete, we wouldn't want to call right away, we would want to wait 5 seconds. We want to guarantee that there is a certain time between requests.

azogue commented:
> This is only really an issue if this object is called every second, which is not something that we allow in Home Assistant.

Well, from my POV this doesn't relate to the update frequency specific value, but to its periodicity. For something with a scan_interval=5, the update is going to be done **each 5 or 6 seconds**, depending on the milliseconds of the utcnow() call, and the milliseconds of the tick. I'm just trying to fix that.

> If the update interval is every 5 seconds (the lowest we allow) and a request takes 5 seconds to complete, we wouldn't want to call right away, we would want to wait 5 seconds. We want to guarantee that there is a certain time between requests.

I agree with guaranteeing some time between calls. But IMO the example is a bad one: if you have a request that takes 5 seconds you have some problem anyway :)

Are you open to the other possibility I pointed? 
Doing a **`utcnow().replace(microsecond=0)+update_interval`** addresses the precision error (as any tick will be on >0 ns) and you'll always have a minimum of `update_interval - 1` between calls.
* For any request that takes < 1s, perfect periodic update each update_interval
* For requests with n.xxx duration, next update in exactly update_interval + n

balloob commented:
`utcnow().replace(microsecond=0)+update_interval` sounds good. Please make sure you add a comment.

azogue committed:
Floor the utcnow timestamp

to remove sub-second error precision, and generate constant frequency updates
as long as the update takes < 1s.
